File: go/algorithms/bitonic.go
package algorithms

// BitonicSort sorts in-place; works best when length is a power of two.
func BitonicSort(arr []int) {
    if len(arr) == 0 {
        return
    }
    bitonicSortRec(arr, 0, len(arr), true)
}

func bitonicSortRec(arr []int, low, cnt int, dir bool) {
    if cnt <= 1 {
        return
    }
    k := cnt / 2
    bitonicSortRec(arr, low, k, true)
    bitonicSortRec(arr, low+k, cnt-k, false)
    bitonicMerge(arr, low, cnt, dir)
}

func bitonicMerge(arr []int, low, cnt int, dir bool) {
    if cnt <= 1 {
        return
    }
    k := cnt / 2
    for i := low; i < low+k && i+k < low+cnt; i++ {
        if (dir && arr[i] > arr[i+k]) || (!dir && arr[i] < arr[i+k]) {
            arr[i], arr[i+k] = arr[i+k], arr[i]
        }
    }
    bitonicMerge(arr, low, k, dir)
    bitonicMerge(arr, low+k, cnt-k, dir)
}

File: rust/src/bitonic.rs
pub fn bitonic_sort(arr: &mut [i32]) {
    let n = arr.len();
    if n == 0 {
        return;
    }
    bitonic_sort_rec(arr, 0, n, true);
}

fn bitonic_sort_rec(arr: &mut [i32], low: usize, cnt: usize, dir: bool) {
    if cnt <= 1 {
        return;
    }
    let k = cnt / 2;
    bitonic_sort_rec(arr, low, k, true);
    bitonic_sort_rec(arr, low + k, cnt - k, false);
    bitonic_merge(arr, low, cnt, dir);
}

fn bitonic_merge(arr: &mut [i32], low: usize, cnt: usize, dir: bool) {
    if cnt <= 1 {
        return;
    }
    let k = cnt / 2;
    for i in low..(low + k) {
        if i + k < low + cnt {
            let should_swap = (dir && arr[i] > arr[i + k]) || (!dir && arr[i] < arr[i + k]);
            if should_swap {
                arr.swap(i, i + k);
            }
        }
    }
    bitonic_merge(arr, low, k, dir);
    bitonic_merge(arr, low + k, cnt - k, dir);
}
